File: src/core/models/artifact.ts
/**
 * Artifact model
 *
 * Represents a shareable artifact in the multi-agent system.
 * Artifacts are used for agent-to-agent communication, such as:
 * - Screenshots (for UI verification)
 * - Test results (for quality gates)
 * - Code diffs (for code review)
 * - Logs (for debugging)
 * - Canvas (for analytical artifacts — fitness reports, dashboards)
 */

export type ArtifactType =
  | "screenshot"
  | "test_results"
  | "code_diff"
  | "logs"
  | "canvas"
  | "attachment";

/**
 * Artifact types that count as agent delivery evidence. `attachment` is
 * user-provided task input and never evidence.
 */
export type EvidenceArtifactType = Exclude<ArtifactType, "attachment">;

/** Artifact types an Agent may create through evidence-write APIs. */
export type AgentWritableArtifactType = "screenshot" | "test_results" | "code_diff" | "logs";

export type ArtifactStatus = "pending" | "provided" | "expired";

export interface Artifact {
  id: string;
  /** Type of the artifact */
  type: ArtifactType;
  /** ID of the task this artifact is associated with */
  taskId: string;
  /** ID of the agent that created/provided this artifact */
  providedByAgentId?: string;
  /** ID of the agent that requested this artifact */
  requestedByAgentId?: string;
  /** ID of the artifact request (if this is a response) */
  requestId?: string;
  /** Content of the artifact (base64 for images, text for others) */
  content?: string;
  /** Optional context or description */
  context?: string;
  /** Current status */
  status: ArtifactStatus;
  /** Workspace ID */
  workspaceId: string;
  /** Created timestamp */
  createdAt: Date;
  /** Updated timestamp */
  updatedAt: Date;
  /** Expiration timestamp (optional) */
  expiresAt?: Date;
  /** Additional metadata */
  metadata?: Record<string, string>;
}

export interface ArtifactRequest {
  id: string;
  /** Agent requesting the artifact */
  fromAgentId: string;
  /** Agent that should provide the artifact */
  toAgentId: string;
  /** Type of artifact requested */
  artifactType: ArtifactType;
  /** Task ID the artifact is for */
  taskId: string;
  /** Optional context for the request */
  context?: string;
  /** Current status */
  status: "pending" | "fulfilled" | "rejected" | "expired";
  /** ID of the artifact that fulfilled this request */
  artifactId?: string;
  /** Workspace ID */
  workspaceId: string;
  /** Created timestamp */
  createdAt: Date;
  /** Updated timestamp */
  updatedAt: Date;
}

export function createArtifact(params: {
  id: string;
  type: ArtifactType;
  taskId: string;
  workspaceId: string;
  providedByAgentId?: string;
  requestedByAgentId?: string;
  requestId?: string;
  content?: string;
  context?: string;
  status?: ArtifactStatus;
  expiresAt?: Date;
  metadata?: Record<string, string>;
}): Artifact {
  const now = new Date();
  return {
    id: params.id,
    type: params.type,
    taskId: params.taskId,
    workspaceId: params.workspaceId,
    providedByAgentId: params.providedByAgentId,
    requestedByAgentId: params.requestedByAgentId,
    requestId: params.requestId,
    content: params.content,
    context: params.context,
    status: params.status ?? "pending",
    createdAt: now,
    updatedAt: now,
    expiresAt: params.expiresAt,
    metadata: params.metadata,
  };
}

export function createArtifactRequest(params: {
  id: string;
  fromAgentId: string;
  toAgentId: string;
  artifactType: ArtifactType;
  taskId: string;
  workspaceId: string;
  context?: string;
}): ArtifactRequest {
  const now = new Date();
  return {
    id: params.id,
    fromAgentId: params.fromAgentId,
    toAgentId: params.toAgentId,
    artifactType: params.artifactType,
    taskId: params.taskId,
    context: params.context,
    status: "pending",
    workspaceId: params.workspaceId,
    createdAt: now,
    updatedAt: now,
  };
}

File: crates/routa-core/src/models/artifact.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum ArtifactType {
    #[serde(rename = "screenshot")]
    Screenshot,
    #[serde(rename = "test_results")]
    TestResults,
    #[serde(rename = "code_diff")]
    CodeDiff,
    #[serde(rename = "logs")]
    Logs,
    #[serde(rename = "canvas")]
    Canvas,
    #[serde(rename = "attachment")]
    Attachment,
}

impl ArtifactType {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Screenshot => "screenshot",
            Self::TestResults => "test_results",
            Self::CodeDiff => "code_diff",
            Self::Logs => "logs",
            Self::Canvas => "canvas",
            Self::Attachment => "attachment",
        }
    }

    #[allow(clippy::should_implement_trait)]
    pub fn from_str(value: &str) -> Option<Self> {
        match value {
            "screenshot" => Some(Self::Screenshot),
            "test_results" => Some(Self::TestResults),
            "code_diff" => Some(Self::CodeDiff),
            "logs" => Some(Self::Logs),
            "canvas" => Some(Self::Canvas),
            "attachment" => Some(Self::Attachment),
            _ => None,
        }
    }

    /// `attachment` is user-provided task input, not agent delivery evidence.
    /// Evidence summaries, transition gates, and agent-writable APIs must exclude it.
    pub fn is_attachment(&self) -> bool {
        matches!(self, Self::Attachment)
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum ArtifactStatus {
    #[serde(rename = "pending")]
    Pending,
    #[serde(rename = "provided")]
    Provided,
    #[serde(rename = "expired")]
    Expired,
}

impl ArtifactStatus {
    pub fn as_str(&self) -> &'static str {
        match self {
            Self::Pending => "pending",
            Self::Provided => "provided",
            Self::Expired => "expired",
        }
    }

    #[allow(clippy::should_implement_trait)]
    pub fn from_str(value: &str) -> Option<Self> {
        match value {
            "pending" => Some(Self::Pending),
            "provided" => Some(Self::Provided),
            "expired" => Some(Self::Expired),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Artifact {
    pub id: String,
    #[serde(rename = "type")]
    pub artifact_type: ArtifactType,
    pub task_id: String,
    pub workspace_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provided_by_agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub requested_by_agent_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub context: Option<String>,
    pub status: ArtifactStatus,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub expires_at: Option<DateTime<Utc>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub metadata: Option<BTreeMap<String, String>>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}
